src/lib/prisma-session-store.ts

```diff
@@ -1,440 +1,467 @@
-     *
-     * @param callback a callback providing either the number of sessions
-     * or an error that occurred
-     */
-    public readonly length = async (
-      callback?: (err?: unknown, length?: number) => void
-    ) => {
-      if (!(await this.validateConnection())) return callback?.();
-
-      // XXX More efficient way? XXX
-
-      try {
-        const sessions = await this.prisma.session.findMany({
-          select: { sid: true }, // Limit what gets sent back; can't be empty.
-        });
+  /**
+   * Destroy the session associated with the given `sid`(s).
+   *
+   * @param sid a single or multiple id(s) to remove data for
+   * @param callback a callback notifying that the session(s) have
+   * been destroyed or that an error occurred
+   */
+  public readonly destroy = async (
+    sid: string | string[],
+    callback?: (err?: unknown) => void
+  ) => {
+    if (!(await this.validateConnection())) return callback?.();
+
+    try {
+      if (Array.isArray(sid)) {
+        await Promise.all(sid.map(async (s) => this.destroy(s, callback)));
+      } else {
+        await this.prisma.session.delete({ where: { sid } });
+      }
+    } catch (e: unknown) {
+      this.logger.warn(
+        `Attempt to destroy non-existent session:${String(sid)} ${String(e)}`
+      );
+      if (callback) defer(callback, e);
+    }
 
-        const itemCount = sessions.length;
-        if (callback) defer(callback, undefined, itemCount);
+    if (callback) defer(callback);
+  };
 
-        return itemCount;
-      } catch (e: unknown) {
-        if (callback) defer(callback, e);
-      }
-    };
+  /**
+   * Attempt to fetch session by the given `sid`.
+   *
+   * @param sid the sid to attempt to fetch
+   * @param callback a function to call with the results
+   */
+  public readonly get = async (
+    sid: string,
+    callback?: (err?: unknown, val?: Express.SessionData) => void
+  ) => {
+    if (!(await this.validateConnection())) return callback?.();
+
+    const session = await this.prisma.session
+      .findOne({
+        where: { sid },
+      })
+      .catch(() => null);
+
+    if (session === null) return callback?.();
+
+    try {
+      const result = this.serializer.parse(
+        session.data ?? '{}'
+      ) as Express.SessionData;
+      if (callback) defer(callback, undefined, result);
+
+      return result;
+    } catch (e: unknown) {
+      this.logger.error(`get(): ${String(e)}`);
+      if (callback) defer(callback, e);
+    }
+  };
 
-    /**
-     * Remove only expired entries from the store
-     */
-    public readonly prune = async () => {
-      if (!(await this.validateConnection())) return;
+  /**
+   * Fetch all sessions' ids
+   *
+   * @param callback a callback providing all session id
+   * or an error that occurred
+   */
+  public readonly ids = async (
+    callback?: (err?: unknown, ids?: number[]) => void
+  ) => {
+    if (!(await this.validateConnection())) return callback?.();
```

Comment:
(Though it is perfectly clear as-is...)

Reply:
The only way I could think of making it shorter would be to throw an error in `validateConnection()`, that would cause the function to end early, however this would also alter the behavior and may cause unexpected results in `express-session` as I believe it does not expect errors to be thrown.

The problem this line solves is that it calls the callback and ends the function. I can't think of any reusable way to end each function early without calling return, which would need to be conditional which makes the if statement required. The await is required because `validateConnection()` needs to use the async Prisma calls and the optional chaining for `callback` is required as it could be undefined. I am pretty sure this is as short as it gets while maintaining the same behavior.

Reply:
Sounds right to me.

Reply:
I did think of a solution, writing a decorator.
It would be relatively easy if these were methods but since they are properties that happen to be functions (as per the `express-session` interface, it makes it much more complicated.

I am happy with this for now.

Reply:
Interesting! 
I keep forgetting that decorators exist (and am still learning what they are used for...)